packages/react/popper/src/Popper.tsx

```diff
@@ -105,71 +126,136 @@ const PopperContent = React.forwardRef<PopperContentElement, PopperContentProps>
+    // assign the reference dynamically once `Content` has mounted so we can collocate the logic
+    useLayoutEffect(() => {
+      reference(context.anchor);
+    }, [reference, context.anchor]);
 
-    const windowSize = useWindowSize();
-    const collisionBoundariesRect = windowSize
-      ? DOMRect.fromRect({ ...windowSize, x: 0, y: 0 })
-      : undefined;
-
-    const { popperStyles, arrowStyles, placedSide, placedAlign } = getPlacementData({
-      anchorRect,
-      popperSize: contentSize,
-      arrowSize,
-
-      // config
-      arrowOffset,
-      side,
-      sideOffset,
-      align,
-      alignOffset,
-      shouldAvoidCollisions: avoidCollisions,
-      collisionBoundariesRect,
-      collisionTolerance,
-    });
-    const isPlaced = placedSide !== undefined;
+    const isPlaced = x !== null && y !== null;
+    const [placedSide, placedAlign] = getSideAndAlignFromPlacement(placement);
+    const { arrow: { x: arrowX, y: arrowY } = {} } = middlewareData;
+    const cannotCenterArrow = middlewareData.arrow?.centerOffset !== 0;
 
     const [contentZIndex, setContentZIndex] = React.useState<string>();
     useLayoutEffect(() => {
       if (content) setContentZIndex(window.getComputedStyle(content).zIndex);
     }, [content]);
 
+    const { hasParent, positionUpdateFns } = usePositionContext(CONTENT_NAME, __scopePopper);
+    const isRoot = !hasParent;
+
+    React.useLayoutEffect(() => {
+      if (!isRoot) {
+        positionUpdateFns.add(update);
+        return () => {
+          positionUpdateFns.delete(update);
+        };
+      }
+    }, [isRoot, positionUpdateFns, update]);
+
+    // when nested contents are rendered in portals, they are appended out of order
+    // we need to re-compute the positioning once the parent has finally been placed.
+    React.useLayoutEffect(() => {
+      if (isRoot && isPlaced) {
+        Array.from(positionUpdateFns)
+          .reverse()
+          .forEach((fn) => requestAnimationFrame(fn));
+      }
+    }, [isRoot, isPlaced, positionUpdateFns]);
+
+    const commonProps = {
+      'data-side': placedSide,
+      'data-align': placedAlign,
+      ...contentProps,
+      ref: composedRefs,
+      style: {
+        ...contentProps.style,
+        // if the PopperContent hasn't been placed yet (not all measurements done)
+        // we prevent animations so that users's animation don't kick in too early referring wrong sides
+        animation: !isPlaced ? 'none' : undefined,
+      },
+    };
+
     return (
-      <div style={{ ...popperStyles, zIndex: contentZIndex }} data-radix-popper-content-wrapper="">
+      <div
+        ref={floating}
+        data-radix-popper-content-wrapper=""
+        style={{
+          position: strategy,
+          left: 0,
+          top: 0,
+          transform: isPlaced
+            ? `translate3d(${Math.round(x)}px, ${Math.round(y)}px, 0)`
+            : 'translate3d(0, -200vh, 0)',
```

Comment:
I've forgotten the purpose of the `-200vh`, do we still need this?

Reply:
This is to make sure it's not overlapping other elements.
See here for more details on its origin: https://github.com/radix-ui/primitives/pull/541